Xpetra/Epetra Undefined Reference when Compiling MueLu Test

@trilinos/muelu @trilinos/xpetra 

Compiling Trilinos (with tests) for KNL on Argonne Theta. I am using the Master branch.

Using:
*  Cray MPI
*  Intel 18.2 Compiler
*  OpenMP enabled

```
[100%] Built target MueLu_Viz
[100%] Built target MueLu_Convergence
[100%] Linking CXX executable MueLu_Driver.exe
Warning:
 Headers and libraries from cray-libsci/18.04.1 will be ignored because they conflict with -mkl.

../../../kokkos/core/src/libkokkoscore.a(Kokkos_Profiling_Interface.cpp.o): In function `Kokkos::Profiling::initialize()':

../../../kokkos/core/src/libkokkoscore.a(Kokkos_Profiling_Interface.cpp.o): In function `Kokkos::Profiling::initialize()':
Kokkos_Profiling_Interface.cpp:(.text+0x5c5): warning: Using 'dlopen' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
CMakeFiles/MueLu_Driver.dir/Driver.cpp.o: In function `int main_<double, int, long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >(Teuchos::CommandLineProcessor&, Xpetra::UnderlyingLib&, int, char**)':
Driver.cpp:(.text._Z5main_IdilN6Kokkos6Compat23KokkosDeviceWrapperNodeINS0_6SerialENS0_9HostSpaceEEEEiRN7Teuchos20CommandLineProcessorERN6Xpetra13UnderlyingLibEiPPc[_Z5main_IdilN6Kokkos6Compat23KokkosDeviceWrapperNodeINS0_6SerialENS0_9HostSpaceEEEEiRN7Teuchos20CommandLineProcessorERN6Xpetra13UnderlyingLibEiPPc]+0x31ca): undefined reference to `Epetra_MultiVector& Xpetra::toEpetra<long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >(Xpetra::MultiVector<double, int, long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >&)'
Driver.cpp:(.text._Z5main_IdilN6Kokkos6Compat23KokkosDeviceWrapperNodeINS0_6SerialENS0_9HostSpaceEEEEiRN7Teuchos20CommandLineProcessorERN6Xpetra13UnderlyingLibEiPPc[_Z5main_IdilN6Kokkos6Compat23KokkosDeviceWrapperNodeINS0_6SerialENS0_9HostSpaceEEEEiRN7Teuchos20CommandLineProcessorERN6Xpetra13UnderlyingLibEiPPc]+0x3213): undefined reference to `Epetra_MultiVector& Xpetra::toEpetra<long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >(Xpetra::MultiVector<double, int, long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >&)'
CMakeFiles/MueLu_Driver.dir/Driver.cpp.o: In function `int main_<double, int, long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> >(Teuchos::CommandLineProcessor&, Xpetra::UnderlyingLib&, int, char**)':
Driver.cpp:(.text._Z5main_IdilN6Kokkos6Compat23KokkosDeviceWrapperNodeINS0_6OpenMPENS0_9HostSpaceEEEEiRN7Teuchos20CommandLineProcessorERN6Xpetra13UnderlyingLibEiPPc[_Z5main_IdilN6Kokkos6Compat23KokkosDeviceWrapperNodeINS0_6OpenMPENS0_9HostSpaceEEEEiRN7Teuchos20CommandLineProcessorERN6Xpetra13UnderlyingLibEiPPc]+0x31ca): undefined reference to `Epetra_MultiVector& Xpetra::toEpetra<long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> >(Xpetra::MultiVector<double, int, long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> >&)'
Driver.cpp:(.text._Z5main_IdilN6Kokkos6Compat23KokkosDeviceWrapperNodeINS0_6OpenMPENS0_9HostSpaceEEEEiRN7Teuchos20CommandLineProcessorERN6Xpetra13UnderlyingLibEiPPc[_Z5main_IdilN6Kokkos6Compat23KokkosDeviceWrapperNodeINS0_6OpenMPENS0_9HostSpaceEEEEiRN7Teuchos20CommandLineProcessorERN6Xpetra13UnderlyingLibEiPPc]+0x3213): undefined reference to `Epetra_MultiVector& Xpetra::toEpetra<long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> >(Xpetra::MultiVector<double, int, long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> >&)'
```

I have the same issue on Arch Linux with the current release version (12.14.1):

` Driver.cpp:(.text._Z5main_IdilN6Kokkos6Compat23KokkosDeviceWrapperNodeINS0_6SerialENS0_9HostSpaceEEEEiRN7Teuchos20CommandLineProcessorERN6Xpetra13UnderlyingLibEiPPc[_Z5main_IdilN6Kokkos6Compat23KokkosDeviceWrapperNodeINS0_6SerialENS0_9HostSpaceEEEEiRN7Teuchos20CommandLineProcessorERN6Xpetra13UnderlyingLibEiPPc]+0x4459): undefined reference to 'Epetra_MultiVector& Xpetra::toEpetra<long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >(Xpetra::MultiVector<double, int, long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >&)'`

my cmake command is very similar

```
     cmake .. -DTrilinos_ENABLE_ALL_OPTIONAL_PACKAGES:BOOL=ON \                                      
              -DTrilinos_ENABLE_ALL_PACKAGES:BOOL=ON \                                               
              -DTrilinos_ENABLE_Gtest:BOOL=ON \                                                      
              -DTrilinos_ENABLE_TESTS=ON \                                                           
              -DTPL_ENABLE_gtest:BOOL=ON \                                                           
              -DTPL_ENABLE_MPI:BOOL=ON \                                                             
              -DTPL_ENABLE_HDF5:BOOL=ON \                                                            
              -DPYTHON_EXECUTABLE:PATH=/usr/bin/python2 \                                            
              -DCMAKE_INSTALL_PREFIX:PATH=/usr \                                                     
              -DBUILD_SHARED_LIBS:BOOL=ON
```

Since

I'm seeing it compiling Trilinos v12.18.1 on CentOS-7 HPC system:

```
Simple.cpp:(.text._Z11SystemSolveIdilN6Kokkos6Compat23KokkosDeviceWrapperNodeINS0_6SerialENS0_9HostSpaceEEEEvRN7Teuchos3RCPIN6Xpetra6MatrixIT_T0_T1_T2_EEEERNS7_INS8_11MultiVectorISA_SB_SC_SD_EEEESK_RNS7_IN5MueLu9HierarchyISA_SB_SC_SD_EEEERNS7_INS8_8OperatorISA_SB_SC_SD_EEEERNS6_18basic_FancyOStreamIcSt11char_traitsIcEEENSt7__cxx1112basic_stringIcSW_SaIcEEES12_bbbiibbid[_Z11SystemSolveIdilN6Kokkos6Compat23KokkosDeviceWrapperNodeINS0_6SerialENS0_9HostSpaceEEEEvRN7Teuchos3RCPIN6Xpetra6MatrixIT_T0_T1_T2_EEEERNS7_INS8_11MultiVectorISA_SB_SC_SD_EEEESK_RNS7_IN5MueLu9HierarchyISA_SB_SC_SD_EEEERNS7_INS8_8OperatorISA_SB_SC_SD_EEEERNS6_18basic_FancyOStreamIcSt11char_traitsIcEEENSt7__cxx1112basic_stringIcSW_SaIcEEES12_bbbiibbid]+0x3b2): undefined reference to `Epetra_MultiVector& Xpetra::toEpetra<long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >(Xpetra::MultiVector<double, int, long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >&)
```

My CMake ends up being pretty complicated, since I have to name the exact location of just about every component, 'lest the configuration find a random, wrong version of various packages, but the components being used are:
```
CMAKE_FLAGS="${CMAKE_FLAGS} \
-DTrilinos_ENABLE_Amesos=ON                      \
-DTrilinos_ENABLE_Epetra=ON                      \
-DTrilinos_ENABLE_EpetraExt=ON                   \
-DTrilinos_ENABLE_Ifpack=ON                      \
-DTrilinos_ENABLE_AztecOO=ON                     \
-DTrilinos_ENABLE_Sacado=ON                      \
-DTrilinos_ENABLE_Teuchos=ON                     \
-DTrilinos_ENABLE_MueLu=ON                       \
-DTrilinos_ENABLE_ML=ON                          \
-DTrilinos_ENABLE_ROL=ON                         \
-DTrilinos_ENABLE_ROL=ON                         \
-DTrilinos_ENABLE_MUMPS=ON                       \
-DTpetra_INST_COMPLEX_DOUBLE=ON                  \
-DTpetra_INST_COMPLEX_FLOAT=ON                   \
-DTrilinos_ENABLE_Zoltan=ON "
#
# and to satisfy some conflicts:
CMAKE_FLAGS="${CMAKE_FLAGS} \
-DTeuchos_ENABLE_COMPLEX:BOOL=ON            \
-DTrilinos_ENABLE_COMPLEX_DOUBLE=ON "
#
# -DTrilinos_ENABLE_Tpetra=ON                      \
# throw these in for good measure:
CMAKE_FLAGS="${CMAKE_FLAGS} -DSEACASAprepro_lib_ENABLE_SEACASExodus=ON -DTrilinos_ENABLE_SEACASAprepro_lib=ON -DTrilinos_ENABLE_SEACASExodus=ON "
#
```

--- a/packages/muelu/src/Interface/MueLu_ParameterListInterpreter_def.hpp
+++ b/packages/muelu/src/Interface/MueLu_ParameterListInterpreter_def.hpp
@@ -1353,14 +1353,17 @@ namespace MueLu {
       TEUCHOS_TEST_FOR_EXCEPTION(partName != "zoltan" && partName != "zoltan2", Exceptions::InvalidArgument,
                                  "Invalid partitioner name: \"" << partName << "\". Valid options: \"zoltan\", \"zoltan2\"");
 
-      bool switched = false;
-      (void)switched;
 #ifndef HAVE_MUELU_ZOLTAN
+      bool switched = false;
       if (partName == "zoltan") {
         this->GetOStream(Warnings0) << "Zoltan interface is not available, trying to switch to Zoltan2" << std::endl;
         partName = "zoltan2";
         switched = true;
       }
+#else
+# ifndef HAVE_MUELU_ZOLTAN
+      bool switched = false;
+# endif
 #endif
 #ifndef HAVE_MUELU_ZOLTAN2
       if (partName == "zoltan2" && !switched) {

--- a/packages/muelu/test/scaling/DriverCore.hpp
+++ b/packages/muelu/test/scaling/DriverCore.hpp
@@ -223,6 +223,40 @@ void PreconditionerSetup(Teuchos::RCP<Xpetra::Matrix<Scalar,LocalOrdinal,GlobalO
 }
 
 
+#if defined(HAVE_MUELU_EPETRA)
+
+// Helper functions for compilation purposes
+template<class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+struct Matvec_Wrapper{
+  static void UnwrapEpetra(Teuchos::RCP<Xpetra::Matrix<Scalar,LocalOrdinal,GlobalOrdinal,Node> >& A,
+                           Teuchos::RCP<Xpetra::MultiVector<Scalar,LocalOrdinal,GlobalOrdinal,Node> >& X,
+                           Teuchos::RCP<Xpetra::MultiVector<Scalar,LocalOrdinal,GlobalOrdinal,Node> >& B,
+                           Teuchos::RCP<const Epetra_CrsMatrix>& Aepetra,
+                           Teuchos::RCP<Epetra_MultiVector>& Xepetra,
+                           Teuchos::RCP<Epetra_MultiVector>& Bepetra) {
+    throw std::runtime_error("Template parameter mismatch");
+  }
+};
+
+
+template<class GlobalOrdinal>
+struct Matvec_Wrapper<double,int,GlobalOrdinal,Kokkos::Compat::KokkosSerialWrapperNode> {
+  static void UnwrapEpetra(Teuchos::RCP<Xpetra::Matrix<double,int,GlobalOrdinal,Kokkos::Compat::KokkosSerialWrapperNode> >& A,
+                           Teuchos::RCP<Xpetra::MultiVector<double,int,GlobalOrdinal,Kokkos::Compat::KokkosSerialWrapperNode> >& X,
+                           Teuchos::RCP<Xpetra::MultiVector<double,int,GlobalOrdinal,Kokkos::Compat::KokkosSerialWrapperNode> >& B,
+                           Teuchos::RCP<const Epetra_CrsMatrix>& Aepetra,
+                           Teuchos::RCP<Epetra_MultiVector>& Xepetra,
+                           Teuchos::RCP<Epetra_MultiVector>& Bepetra) {
+    typedef double SC;
+    typedef int LO;
+    typedef GlobalOrdinal GO;
+    typedef Kokkos::Compat::KokkosSerialWrapperNode NO;
+    Aepetra = Xpetra::Helpers<SC, LO, GO, NO>::Op2EpetraCrs(A);
+    Xepetra = Teuchos::rcp(& Xpetra::toEpetra(*X),false);
+    Bepetra = Teuchos::rcp(& Xpetra::toEpetra(*B),false);
+  }
+};
+#endif
 
 //*************************************************************************************
 template<class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
@@ -271,13 +305,11 @@ void SystemSolve(Teuchos::RCP<Xpetra::Matrix<Scalar,LocalOrdinal,GlobalOrdinal,N
         Btpetra = rcp(& Xpetra::toTpetra(*B),false);
       }
 #endif
-#if defined(HAVE_MUELU_EPETRA) && !defined(HAVE_MUELU_INST_COMPLEX_INT_INT) && !defined(HAVE_MUELU_INST_FLOAT_INT_INT)
+#if defined(HAVE_MUELU_EPETRA)
       Teuchos::RCP<const Epetra_CrsMatrix> Aepetra;
       Teuchos::RCP<Epetra_MultiVector> Xepetra,Bepetra;
       if(lib==Xpetra::UseEpetra) {
-        Aepetra = Xpetra::Helpers<SC, LO, GO, NO>::Op2EpetraCrs(A);
-        Xepetra = rcp(& Xpetra::toEpetra(*X),false);
-        Bepetra = rcp(& Xpetra::toEpetra(*B),false);
+        Matvec_Wrapper<SC,LO,GO,NO>::UnwrapEpetra(A,X,B,Aepetra,Xepetra,Bepetra);
       }
 #endif
 

--- a/packages/xpetra/sup/Utils/Xpetra_MatrixUtils.hpp
+++ b/packages/xpetra/sup/Utils/Xpetra_MatrixUtils.hpp
@@ -453,7 +453,7 @@ public:
     Ac->getLocalDiagCopy(*diagVec);
 
     LocalOrdinal lZeroDiags = 0;
-    Teuchos::ArrayRCP< Scalar > diagVal = diagVec->getDataNonConst(0);
+    Teuchos::ArrayRCP< const Scalar > diagVal = diagVec->getData(0);
 
     for (size_t i = 0; i < rowMap->getNodeNumElements(); i++) {
       if (TST::magnitude(diagVal[i]) <= threshold) {
@@ -521,7 +521,7 @@ public:
 #ifdef HAVE_XPETRA_DEBUG // only for debugging
     // check whether Ac has been repaired...
     Ac->getLocalDiagCopy(*diagVec);
-    Teuchos::ArrayRCP< Scalar > diagVal2 = diagVec->getDataNonConst(0);
+    diagVal = diagVec->getData(0);
     for (size_t r = 0; r < Ac->getRowMap()->getNodeNumElements(); r++) {
       if (TST::magnitude(diagVal[r]) <= threshold) {
         fos << "Error: there are too small entries left on diagonal after repair..." << std::endl;

